File: problems/Python/52. N-Queens II.py
class Solution:
    def totalNQueens(self, n: int) -> int:
        cols = set()
        pdia = set()
        ndia = set()

        def bt(r):
            if r == n:
                return 1
            
            cnt = 0
            for c in range(n):
                if c in cols or (r - c) in pdia or (r + c) in ndia:
                    continue
                
                cols.add(c)
                pdia.add(r - c)
                ndia.add(r + c)
                
                cnt += bt(r + 1)
                
                cols.remove(c)
                pdia.remove(r - c)
                ndia.remove(r + c)
            return cnt

        return bt(0)

File: problems/Java/52. N-Queens II.java
class Solution {
    public int totalNQueens(int n) {
        return bt(0, n, new HashSet<>(), new HashSet<>(), new HashSet<>());
    }
    
    private int bt(int r, int n, Set<Integer> cols, Set<Integer> pdia, Set<Integer> ndia) {
        if (r == n) {
            return 1;
        }

        int cnt = 0;
        for (int c = 0; c < n; c++) {
            if (cols.contains(c) || pdia.contains(r - c) || ndia.contains(r + c)) {
                continue;
            }

            cols.add(c);
            pdia.add(r - c);
            ndia.add(r + c);
            
            cnt += bt(r + 1, n, cols, pdia, ndia);

            cols.remove(c);
            pdia.remove(r - c);
            ndia.remove(r + c);
        }
        return cnt;
    }
}